File: src/Haskell/X86Prg.hs
{-# LANGUAGE MultiParamTypeClasses #-}
module X86Prg (
  X86Prg(..)
  ) where

import Backend
import X86Instr
import X86Function

data X86Prg = X86Prg { x86Functions :: [X86Function] }

instance MachinePrg X86Prg X86Function X86Instr where
  machinePrgFunctions = x86Functions
  replaceFunctions p m = p { x86Functions = m }

instance Show X86Prg where
  show prg = error "TODO"

File: src/Rust/x86/x86prg.rs
use backend::MachinePrg;
use x86::X86Function;
use x86::X86Instr;

#[derive(Debug)]
pub struct X86Prg {
    pub functions: Vec<X86Function>,
}

impl MachinePrg<X86Instr, X86Function> for X86Prg {
    fn functions(&self) -> &Vec<X86Function> {
        &self.functions
    }

    fn functions_mut(&mut self) -> &mut Vec<X86Function> {
        &mut self.functions
    }
}
